Move complex extension test to c10

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28468 Delete ATenDispatch
* #28186 Remove c10->ATen registration forwarding
* #27978 Migrate remaining ops to the c10 dispatcher
* **#28208 Move complex extension test to c10**
* #28184 Fix codegen for out operators
* #28183 Allow migrating factory methods to c10
* #28182 Fix overload names
* #28181 Add unsupported types to schema type parser

Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

smessmer committed:
Move complex extension test to c10

Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]

smessmer committed:
Update on "Move complex extension test to c10"


Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]

smessmer committed:
Update on "Move complex extension test to c10"


Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]

smessmer committed:
Update on "Move complex extension test to c10"


Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]

smessmer committed:
Update on "Move complex extension test to c10"


Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]

ezyang commented on test/cpp_extensions/complex_registration_extension.cpp:
```diff
@@ -46,10 +47,11 @@ static Tensor empty_complex(IntArrayRef size, const TensorOptions & options, c10
+  .op(torch::RegisterOperators::options()
+    .schema("aten::empty.memory_format(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor")
+    .impl_unboxedOnlyKernel<decltype(empty_complex), &empty_complex>(TensorTypeId::ComplexCPUTensorId)
+    .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA));
```
The new API seems worse than the old one. As a pre-existing schema string that we defined, the custom registration shouldn't be in the business of specifying the alias analysis: the aliasing was predetermined by us, the library authors, user implementations of the contract don't get to change the contract.

I have also mentioned in 1:1 that I don't really want the code to be explicitly mentioning ComplexCPUTensorId, though this is a defect of the old code too.

smessmer replied:
Both good points. Unrelated to this PR though. XLA and everything else is already using this new API, complex was somehow missed. We should fix the API but not block this PR by it.

smessmer committed:
Update on "Move complex extension test to c10"


Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]

smessmer committed:
Update on "Move complex extension test to c10"


Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]

smessmer committed:
Update on "Move complex extension test to c10"


Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]

smessmer committed:
Update on "Move complex extension test to c10"


Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]

smessmer committed:
Update on "Move complex extension test to c10"


Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]

smessmer committed:
Update on "Move complex extension test to c10"


Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]

smessmer committed:
Update on "Move complex extension test to c10"


Backend extensions should call torch::RegisterOperators, not globalATenDispatch().
If the op is still on globalATenDispatch, then torch::RegisterOperators will do the right thing and forward it to globalATenDispatch.

Differential Revision: [D17975369](https://our.internmc.facebook.com/intern/diff/D17975369/)

[ghstack-poisoned]
